File: modules/frontend/src/lib/root/project-catalog.ts
/**
 * The attached catalog as a new thread needs it: which project you were last
 * working in, and what to call it in the space of one word.
 *
 * Nothing here draws anything. The selector above it is only about presenting
 * a choice, and the order that choice is offered in is a fact about the
 * workspace rather than a fact about the menu.
 */

import type { Project, ThreadListItem } from "@artisan/protocol";
import { ThreadLastMessageAt } from "./thread-navigation";

/** One project as the selector needs it: an identity, a place in time, its weight. */
export type RecentProject = {
	/**
	 * When a message was last sent here. A project's own `updated_at` is a catalog
	 * row's stamp — it moves when the folder is attached or renamed — so the
	 * threads inside it are the better witness. The row is only the fallback for
	 * a project that has no thread messages yet.
	 */
	readonly last_message_at: string;
	readonly project: Project;
	readonly thread_count: number;
};

/**
 * Freshest first, which is the only order a project list means anything in:
 * the one you want next is nearly always the one you had last.
 */
export const RecentProjects = (
	projects: ReadonlyArray<Project>,
	threads: ReadonlyArray<ThreadListItem>,
): ReadonlyArray<RecentProject> => {
	const activity = new Map<string, { last_message_at: string; thread_count: number }>();
	for (const thread of threads) {
		const project_id = thread.primary_project?.project_id;
		if (project_id === undefined) continue;
		const last_message_at = ThreadLastMessageAt(thread);
		const seen = activity.get(project_id);
		if (seen === undefined) {
			activity.set(project_id, {
				last_message_at,
				thread_count: 1,
			});
			continue;
		}
		seen.thread_count += 1;
		if (last_message_at.localeCompare(seen.last_message_at) > 0)
			seen.last_message_at = last_message_at;
	}

	return [...projects]
		.map((project): RecentProject => {
			const seen = activity.get(project.project_id);
			return {
				last_message_at: seen?.last_message_at ?? project.updated_at,
				project,
				thread_count: seen?.thread_count ?? 0,
			};
		})
		.sort((left, right) => right.last_message_at.localeCompare(left.last_message_at));
};

/**
 * The project a fresh surface should open on: the one the reader named, else
 * the one they were last in, else simply the most recent. Resolved here rather
 * than in the route so "which project" has one answer wherever it is asked.
 */
export const PreferredProject = (
	recents: ReadonlyArray<RecentProject>,
	preferred_project_id: string | undefined,
): Project | undefined =>
	(preferred_project_id === undefined
		? undefined
		: recents.find((recent) => recent.project.project_id === preferred_project_id)?.project) ??
	recents[0]?.project;

/** Two letters is the most a small square can hold and still be read. */
export const ProjectMonogram = (name: string): string => {
	const parts = name.split(/[-_. /\\]/u).filter((part) => part.length > 0);
	if (parts.length === 0) return "··";
	return parts.length === 1
		? (parts[0] ?? name).slice(0, 2).toUpperCase()
		: `${parts[0]?.[0] ?? ""}${parts[1]?.[0] ?? ""}`.toUpperCase();
};

File: modules/frontend/src/project_catalog.rs
//! Pure recent-project, preferred-project, and monogram policy.
//!
//! This is the dependency-free Rust counterpart of
//! `modules/frontend/src/lib/root/project-catalog.ts`. The types below are
//! deliberately small projections of the protocol rows: catalog ordering
//! needs only a project's identity and update stamp, while thread activity
//! needs its creation stamp, optional message stamp, and optional primary
//! project identity. Protocol conversion, repositories, and rendering stay
//! outside this leaf.

use std::collections::HashMap;

/// The project fields consumed by the recent/preferred catalog policy.
///
/// Other project metadata, such as a display name or root path, is not
/// inspected by this policy and belongs to the caller's full project model.
#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub struct Project {
    /// Stable project identity used to join projects with assigned threads.
    pub project_id: String,
    /// Catalog-row fallback activity when no assigned thread exists.
    pub updated_at: String,
}

impl Project {
    /// Builds a policy projection from its identity and update stamp.
    #[must_use]
    pub fn new(project_id: impl Into<String>, updated_at: impl Into<String>) -> Self {
        Self {
            project_id: project_id.into(),
            updated_at: updated_at.into(),
        }
    }
}

/// The primary-project reference optionally carried by a thread row.
#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub struct ProjectRef {
    /// Identity of the project's primary association.
    pub project_id: String,
}

impl ProjectRef {
    /// Builds a primary-project reference from its identity.
    #[must_use]
    pub fn new(project_id: impl Into<String>) -> Self {
        Self {
            project_id: project_id.into(),
        }
    }
}

/// The thread fields consumed by the recent-project policy.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ThreadRow {
    /// Creation stamp used when no last-message stamp is present.
    pub created_at: String,
    /// Last-message stamp, when the thread has one.
    pub last_message_at: Option<String>,
    /// Optional primary project assignment; unassigned rows do not count.
    pub primary_project: Option<ProjectRef>,
}

impl ThreadRow {
    /// Builds a thread policy projection.
    #[must_use]
    pub fn new(
        created_at: impl Into<String>,
        last_message_at: Option<String>,
        primary_project: Option<ProjectRef>,
    ) -> Self {
        Self {
            created_at: created_at.into(),
            last_message_at,
            primary_project,
        }
    }

    /// Returns the stamp used as this row's project activity.
    #[must_use]
    pub fn activity_timestamp(&self) -> &str {
        self.last_message_at.as_deref().unwrap_or(&self.created_at)
    }
}

/// One project enriched with the activity used to order the catalog.
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct RecentProject {
    /// Newest assigned-thread activity, or the project's update stamp when
    /// no assigned activity exists.
    pub last_message_at: String,
    /// The project represented by this result.
    pub project: Project,
    /// Number of assigned thread rows for this project.
    pub thread_count: usize,
}

#[derive(Debug)]
struct ProjectActivity {
    last_message_at: String,
    thread_count: usize,
}

/// Builds a newest-first project catalog from project and thread rows.
///
/// A thread contributes only when it has a primary project reference. Its
/// activity stamp is `last_message_at` when present and `created_at`
/// otherwise. Counts include every assigned row whose project id matches,
/// while the activity stamp keeps the lexicographically newest ISO timestamp
/// for that id. Projects with no matching activity use `updated_at` and have
/// a zero count.
///
/// The final `sort_by` is stable. Equal timestamps therefore retain the
/// caller's project order, which makes ties deterministic without inventing a
/// second ordering rule that the TypeScript policy does not have.
#[must_use]
pub fn recent_projects(projects: &[Project], threads: &[ThreadRow]) -> Vec<RecentProject> {
    let mut activity = HashMap::<String, ProjectActivity>::new();

    for thread in threads {
        let Some(project) = thread.primary_project.as_ref() else {
            continue;
        };

        let timestamp = thread.activity_timestamp();
        let seen = activity
            .entry(project.project_id.clone())
            .or_insert_with(|| ProjectActivity {
                last_message_at: timestamp.to_owned(),
                thread_count: 0,
            });
        seen.thread_count += 1;
        if timestamp > seen.last_message_at.as_str() {
            timestamp.clone_into(&mut seen.last_message_at);
        }
    }

    let mut recents = projects
        .iter()
        .map(|project| {
            let seen = activity.get(&project.project_id);
            RecentProject {
                last_message_at: seen.map_or_else(
                    || project.updated_at.clone(),
                    |item| item.last_message_at.clone(),
                ),
                project: project.clone(),
                thread_count: seen.map_or(0, |item| item.thread_count),
            }
        })
        .collect::<Vec<_>>();

    recents.sort_by(|left, right| right.last_message_at.cmp(&left.last_message_at));
    recents
}

/// Chooses the requested recent project, falling back to the first recent
/// project and then to no project when the catalog is empty.
#[must_use]
pub fn preferred_project<'a>(
    recents: &'a [RecentProject],
    preferred_project_id: Option<&str>,
) -> Option<&'a Project> {
    preferred_project_id
        .and_then(|project_id| {
            recents
                .iter()
                .find(|recent| recent.project.project_id == project_id)
                .map(|recent| &recent.project)
        })
        .or_else(|| recents.first().map(|recent| &recent.project))
}

/// Returns the preferred project as an owned value for callers that need to
/// retain it after the recent-result slice is released.
#[must_use]
pub fn preferred_project_owned(
    recents: &[RecentProject],
    preferred_project_id: Option<&str>,
) -> Option<Project> {
    preferred_project(recents, preferred_project_id).cloned()
}

/// Placeholder shown when a project name contains no usable parts.
pub const EMPTY_MONOGRAM: &str = "··";

/// Builds the compact project monogram used by a small project tile.
///
/// Names split on hyphen, underscore, period, ASCII space, forward slash, or
/// backslash. Empty parts are discarded. A single remaining part contributes
/// its first two Unicode scalar values; multiple parts contribute the first
/// scalar of each of the first two parts. Each selected scalar is uppercased
/// with Rust's Unicode mapping, including mappings that expand to more than
/// one scalar.
#[must_use]
pub fn project_monogram(name: &str) -> String {
    let parts = name
        .split(['-', '_', '.', ' ', '/', '\\'])
        .filter(|part| !part.is_empty())
        .collect::<Vec<_>>();

    match parts.as_slice() {
        [] => EMPTY_MONOGRAM.to_owned(),
        [part] => uppercase_first_scalars(part, 2),
        [first, second, ..] => {
            let mut monogram = uppercase_first_scalars(first, 1);
            monogram.push_str(&uppercase_first_scalars(second, 1));
            monogram
        }
    }
}

fn uppercase_first_scalars(value: &str, count: usize) -> String {
    let mut uppercase = String::new();
    for character in value.chars().take(count) {
        uppercase.extend(character.to_uppercase());
    }
    uppercase
}
